akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -137,6 +143,91 @@ object GraphStageLogic {
   private object DoNothing extends (() ⇒ Unit) {
     def apply(): Unit = ()
   }
+
+  trait MinimalDeathWatch {
+    this: MinimalActorRef ⇒
+    private var watchedBy = ActorCell.emptyActorRefSet
```

Comment:
yes, this should be an AtomicReference and either contain a list of watchers or a tombstone. `sendTerminated` should do a `getAndSet` to install the tombstone and notify watchers, while addWatcher should do `compareAndSet` to add watchers, or simply send `DeathWatchNotification` if there is a tombstone already.